tty: Deno.setRaw(rid, mode) to turn on/off raw mode

kevinkassimo opened the pull request:
Difference from the first option:
+ `Deno.setRaw(rid, mode)` to turn on/off raw mode
+ Added `Deno.isatty(rid)` that can test any resources, not just stdin/stdout/stderr (personally prefer deprecating current `Deno.isTTY()`)

```ts
Deno.setRaw(0, true);
Deno.setRaw(0, true); // safe

const buf = new Uint8Array(6);
for (let i = 0; i < 6; i++) {
  const nread = await Deno.stdin.read(buf);
  if (nread === Deno.EOF) {
    break;
  } else {
    console.log(`READ ${nread} bytes:`, new TextDecoder().decode(buf.subarray(0, nread)));
  }
}

Deno.setRaw(0, false);
Deno.setRaw(0, false); // safe
```

ry commented on cli/js/lib.deno.ns.d.ts:
```diff
@@ -543,6 +537,24 @@ declare namespace Deno {
+  /** An instance of `File` for stdout. */
+  export const stdout: File;
+  /** An instance of `File` for stderr. */
+  export const stderr: File;
```
I like the minimality of `isatty(rid)`. Maybe you can do something similar here, instead of defining this new class `TTYInput` you could just add a new function... something like `Deno.setRaw(rid)`. What do you think?

Additionally can you mark in the JSDOC any new APIs as 
```
/** UNSTABLE: newly added API
  *
  *  Check if a given resource is TTY
  */
export function isatty(rid: number): boolean;
```

kevinkassimo replied:
Do you think the original returning a restoration function after calling `Deno.setRaw(rid)` works here? (In case I want to store the information for restoring without exposing? This would be a pure TS side change, and similar to what I did in the old PR, but without wrapping in a class)

ry replied:
I don't think storing the restore state is so important. The user can do that.

kevinkassimo replied:
I don't think they can with current API (unless we expose a full representation `<termios>`) if we do not store for them? And there is the discrepancy between Unix and Windows here

Node does not provide the API for direct restoration either. I believe with `tty.ReadStream::setRaw(false)` libuv simply [sets back to the saved original mode](https://github.com/libuv/libuv/blob/1399d851b112ad8376ef2a46303e72fd55125439/src/unix/tty.c#L243)

ry replied:
> libuv simply sets back to the saved original mode

Yes we should do that too...

kevinkassimo replied:
The slight annoyance is that we will need to accompany StreamResource (including `FsFile`) with such optional data. I'll see where would be the best place for me to place this piece of info.

kevinkassimo commented:
Added a basic test (unix only, I have no access to Windows). Added `pty` crate as a dev dependency.

`permissions::tests::test_permissions_request_read` and `permissions::tests::test_permission_request_net` seems to be badly flaky

ry commented:
@kevinkassimo I'm a little confused because there is still `Deno.isTTY()` which if I'm reading your patch correctly does not call the new `Deno.isatty()` .. which I would expect it to.

kevinkassimo commented:
@ry I can probably update `isTTY` to use `isatty`. I was originally thinking about deprecating `isTTY` since the 2 has very similar names while `isTTY` only handles stdin/stdout/stderr

ry commented on cli/ops/os.rs:
```diff
@@ -151,18 +149,6 @@ fn op_exit(
-    "stdout": atty::is(atty::Stream::Stdout),
-    "stderr": atty::is(atty::Stream::Stderr),
-  })))
-}
```
Good

ry commented on cli/js/tty.ts:
```diff
@@ -13,3 +13,15 @@ export function setRaw(rid: number, mode: boolean): void {
+    stdout: isatty(1),
+    stderr: isatty(2)
+  };
+}
```
I think we should just remove it actually.

kevinkassimo commented:
Is this PR considered ready? Have been resolving conflicts with master updates for quite a few times 😅
